doc/memos/rdm-draft-petry-riot-design-goals.md

```diff
@@ -0,0 +1,345 @@
+them and choose between different stacks; they should be able to get the best
+performance and most relevant functionality out of whatever resources they have
+available.
+
+## Short learning curve
+
+RIOT's use cases involve makers, researchers in non-computer science fields,
+broadly skilled engineers making proofs-of-concept in IoT startups, and
+experienced embedded C developers.
+
+All of these users should be comfortable using RIOT and its tooling, whatever
+their platform or hardware. No user of any skill level should be tripped up
+because they haven't configured something that they wouldn't reasonably be
+expected to know about. Whatever the user's background, RIOT should demand as
+little RIOT-specific learning as possible. It should, therefore, adhere to
+common systems and networking standards such as C99, POSIX, libc, UDP, CoAP,
+IPv6, and so on wherever it makes sense.
+
+## Vendor and technology independence
+
+Vendors and technologies are supported with zero bias. This means users can
+choose what's best for them, without being led by RIOT. Moreover, RIOT is Free
+Software [2], which means users are free to use it as they wish without
+lock-in.
+
+The choice of the hardware and standards RIOT supports should depend only on
+how well they support the use cases. Any allegiance that modules may have
+should be kept to themselves.
+
+## Versatility
+
+The use cases for IoT systems are monifold, and so are their requirements.
+Therefore, design decisions in RIOT should not prefer one technology or one
+protocol over another. 
+
+Default configurations should support as many users and use cases as possible.
+RIOT should provide everything users need, either in its code base or in the
+external packages it supports. The list of hardware, algorithms and other
+elements that RIOT supports should constantly be expanding.
+
+## Modularity
+
+RIOT decomposes into small modules. Being modular lets RIOT address many
+different memory, functionality, and performance demands. It also helps
+development efforts to scale in a widely distributed community.
+
+Modules should be abstracted from one another as cleanly as possible. It should
+be easy for users to manage or exploit this modularity, for example through
+configuration methods, easy integration of third-party source code, or
+different levels of modularity. The granularity of modules should be stable and
+similar across different types of module. Users shouldn't want to split
+modules, but they shouldn't be unnecessarily fine-grained either.
+
+## Cross-hardware portability
+
+Users might want to write a program for one piece of hardware, and later run it
+on another. RIOT should let user code be completely portable, so long as it's
+valid with the hardware.
+
+The hardware abstraction layer should be stable, well defined and consistent.
+Above the HAL, the only thing that modules should know about hardware is
+whether a required feature is there or not. If it isn't, the module should
```

Comment:
This seems a bit too abstract: "Required features" could be peripherals such as a sensor or actuator, but at the same time a property of the peripheral such as "precision", "speed of operation", etc. I would suggest to differentiate a bit.

Reply:
This is meant to suggest that the modules should only know about whether their dependencies are compiled for that hardware. Perhaps "whether its build dependencies are provided or not"?